### PR TITLE
ARROW-15641: [C++][Python] UDF Aggregate Function Implementation

### PR DESCRIPTION
**Pull request by vibhatha**
Initial Draft of Aggregate UDFs for Python

- [x] Improve the existing logic with more tests
- [ ] Find out a missing aggregate function and try to apply it with the current API (check if the applied theory sustains)
- [x] Using a custom object for state holding rather than Arrow data structures
- [x] Improvement function docs and comments
- [ ] Adding binary and other arity tests
- [x] Self-review 1
- [ ] Self-review 2

### CODE REVIEW

**Comment by vibhatha** [comment]
cc @westonpace appreciate an initial review on the draft PR.

File: python/pyarrow/tests/test_udf.py [hunk 1]
```diff
@@ -504,3 +504,132 @@ def test_input_lifetime(unary_func_fixture):
...
+        def __init__(self, non_null=0):
+            self._non_null = non_null
+
+        @property
+        def non_null(self):
+            return self._non_null
+
+        @non_null.setter
+        def non_null(self, value):
+            self._non_null = value
```
Comment (westonpace) [inline python/pyarrow/tests/test_udf.py:520]: I'm not much of a python expert but getters and setters seem like overkill here.  Are they needed?

File: python/pyarrow/tests/test_udf.py [hunk 2]
```diff
@@ -504,3 +504,132 @@ def test_input_lifetime(unary_func_fixture):
...
+
+        def __repr__(self):
+            if self._non_null is None:
+                return "no values stored"
+            else:
+                return "count: " + str(self.non_null)
+
+    def init():
+        state = State(0)
+        return state
```
Comment (westonpace) [inline python/pyarrow/tests/test_udf.py:530]:
```suggestion
        return State(0)
```

File: python/pyarrow/tests/test_udf.py [hunk 3]
```diff
@@ -504,3 +504,132 @@ def test_input_lifetime(unary_func_fixture):
...
+
+    def consume(ctx, x):
+        if isinstance(x, pa.Array):
+            non_null = pc.sum(pc.invert(pc.is_nan(x))).as_py()
+        elif isinstance(x, pa.Scalar):
```
Comment (westonpace) [inline python/pyarrow/tests/test_udf.py:535]: Can a unary aggregate ever be called with a scalar?

File: python/pyarrow/tests/test_udf.py [hunk 4]
```diff
@@ -504,3 +504,132 @@ def test_input_lifetime(unary_func_fixture):
...
+        return State(non_null)
+
+    def merge(ctx, other_state):
+        merged_state_val = ctx.state.non_null + other_state.non_null
+        return State(merged_state_val)
+
+    def finalize(ctx):
+        return pa.array([ctx.state.non_null])
+
+    func_name = "simple_count"
```
Comment (westonpace) [inline python/pyarrow/tests/test_udf.py:548]: Maybe `valid_count` or `non_null_count`?

File: python/pyarrow/tests/test_udf.py [hunk 5]
```diff
@@ -504,3 +504,132 @@ def test_input_lifetime(unary_func_fixture):
...
+        def non_null(self, value):
+            self._non_null = value
+
+        @property
+        def null(self):
+            return self._null
+
+        @null.setter
+        def null(self, value):
+            self._null = value
```
Comment (westonpace) [inline python/pyarrow/tests/test_udf.py:585]: Same comment on getters and setters

File: python/pyarrow/tests/test_udf.py [hunk 6]
```diff
@@ -504,3 +504,132 @@ def test_input_lifetime(unary_func_fixture):
...
+        def __repr__(self):
+            if self._non_null is None:
+                return "no values stored"
+            else:
+                return "non_null: " + str(self.non_null) \
+                    + ", null: " + str(self.null)
+
+    def init():
+        state = State(0, 0)
+        return state
```
Comment (westonpace) [inline python/pyarrow/tests/test_udf.py:596]:
```suggestion
        return State(0, 0)
```

File: python/pyarrow/tests/test_udf.py [hunk 7]
```diff
@@ -504,3 +504,132 @@ def test_input_lifetime(unary_func_fixture):
...
+        non_null = 0
+        if isinstance(x, pa.Array):
+            non_null = pc.sum(pc.invert(pc.is_nan(x))).as_py()
+            null = len(x) - non_null
+        elif isinstance(x, pa.Scalar):
+            if x.as_py():
+                non_null = 1
+            else:
+                null = 1
+        non_null = non_null + ctx.state.non_null
```
Comment (westonpace) [inline python/pyarrow/tests/test_udf.py:609]:
```suggestion
        non_null = non_null + ctx.state.non_null
        null = null + ctx.state.null
```

File: python/pyarrow/tests/test_udf.py [hunk 8]
```diff
@@ -504,3 +504,132 @@ def test_input_lifetime(unary_func_fixture):
...
+        return State(non_null, null)
+
+    def merge(ctx, other_state):
+        merged_st_non_null = ctx.state.non_null + other_state.non_null
+        merged_st_null = ctx.state.null + other_state.null
+        return State(merged_st_non_null, merged_st_null)
+
+    def finalize(ctx):
+        print(ctx.state)
+        return pa.array([ctx.state.non_null, ctx.state.null])
```
Comment (westonpace) [inline python/pyarrow/tests/test_udf.py:619]: An aggregate UDF must always return an array of length 1.  Do you perhaps want something like...

```
return pa.array([{"non_null": ctx.state.non_null, "null": ctx.state.null}])
```

File: cpp/examples/arrow/udf_example.cc [hunk 9]
```diff
@@ -83,15 +83,154 @@ arrow::Status Execute() {
...
+
+std::unique_ptr<cp::FunctionOptions> SimpleCountFunctionOptionsType::Copy(
+    const cp::FunctionOptions&) const {
+  return std::make_unique<SimpleCountOptions>();
+}
+
+const cp::FunctionDoc simple_count_doc{
+    "SimpleCount the number of null / non-null values",
+    ("By default, only non-null values are counted.\n"
+     "This can be changed through SimpleCountOptions."),
```
Comment (westonpace) [inline cpp/examples/arrow/udf_example.cc:129]: Well, this isn't true quite yet :)

File: python/pyarrow/src/arrow/python/udf.cc [hunk 10]
```diff
@@ -120,6 +128,218 @@ Status RegisterScalarFunction(PyObject* user_function, ScalarUdfWrapperCallback
...
+      OwnedRef result(init_cb(init_function->obj()));
+      PyObject* init_res = result.obj();
+      Py_INCREF(init_res);
+      this->udf_context_ = ScalarAggregateUdfContext{default_memory_pool(), std::move(init_res)};
+      this->owned_state_.reset(result.obj());
+      RETURN_NOT_OK(CheckPyError());
+      return Status::OK();
+    });
+    if (!st.ok()) {
+      throw std::runtime_error(st.ToString());
```
Comment (westonpace) [inline python/pyarrow/src/arrow/python/udf.cc:211]: Is throwing an error the right thing to do?

File: python/pyarrow/src/arrow/python/udf.cc [hunk 11]
```diff
@@ -120,6 +128,218 @@ Status RegisterScalarFunction(PyObject* user_function, ScalarUdfWrapperCallback
...
+      consume_function->detach();
+      merge_function->detach();
+      finalize_function->detach();
+    }
+  }
+
+  void Init(ScalarAggregateInitUdfWrapperCallback& init_cb , std::shared_ptr<OwnedRefNoGIL>& init_function) {
+    auto st =  SafeCallIntoPython([&]() -> Status { 
+      OwnedRef result(init_cb(init_function->obj()));
+      PyObject* init_res = result.obj();
```
Comment (westonpace) [inline python/pyarrow/src/arrow/python/udf.cc:203]: Probably need to do sanity checking (e.g. not null, etc.) on things returned from user callbacks.

File: python/pyarrow/_compute.pyx [hunk 12]
```diff
@@ -2641,3 +2722,200 @@ def register_scalar_function(func, function_name, function_doc, in_types,
...
+
+
+def register_scalar_aggregate_function(init_func, consume_func, merge_func, finalize_func,
+                                       function_name, function_doc, in_types, out_type):
+    """
+    Register a user-defined scalar aggregate function.
+
+    A scalar aggregate function is a set of 4 functions which formulates
+    the operation pieces of an scalar aggregation. The base behavior in 
+    terms of computation is very much similar to scalar functions.
```
Comment (westonpace) [inline python/pyarrow/_compute.pyx:2734]:
```suggestion
    A scalar aggregate function is a set of 4 functions which are
    called at different times during the calculation of the aggregate.
```

I think we are missing a basic description first.  Something like:

```
An aggregate function reduces a column of values into a single aggregate result.
```

File: python/pyarrow/_compute.pyx [hunk 13]
```diff
@@ -2641,3 +2722,200 @@ def register_scalar_function(func, function_name, function_doc, in_types,
 
     check_status(RegisterScalarFunction(c_function,
                                         <function[CallbackUdf]> &_scalar_udf_callback, c_options))
+
+
+def register_scalar_aggregate_function(init_func, consume_func, merge_func, finalize_func,
```
Comment (westonpace) [inline python/pyarrow/_compute.pyx:2727]: Why "scalar aggregate"?  Is there a non-scalar aggregate?

File: python/pyarrow/_compute.pyx [hunk 14]
```diff
@@ -2641,3 +2722,200 @@ def register_scalar_function(func, function_name, function_doc, in_types,
...
+    """
+    Register a user-defined scalar aggregate function.
+
+    A scalar aggregate function is a set of 4 functions which formulates
+    the operation pieces of an scalar aggregation. The base behavior in 
+    terms of computation is very much similar to scalar functions.
+
+    Parameters
+    ----------
+    init_func : callable
```
Comment (westonpace) [inline python/pyarrow/_compute.pyx:2738]: Is `callable` the right word for python?  In python docs it refers to this kind of thing as `function` I think: https://docs.python.org/3/library/functions.html#map

File: python/pyarrow/_compute.pyx [hunk 15]
```diff
@@ -2641,3 +2722,200 @@ def register_scalar_function(func, function_name, function_doc, in_types,
...
+        A callable implementing the user-defined initialization function.
+        This function is used to set the state for the aggregate operation
```
Comment (westonpace) [inline python/pyarrow/_compute.pyx:2740]: I'm not sure that "set the state" is correct.  Maybe "create the initial state"?

File: python/pyarrow/_compute.pyx [hunk 16]
```diff
@@ -2641,3 +2722,200 @@ def register_scalar_function(func, function_name, function_doc, in_types,
...
+        and returns the state object.
+    consume_func : callable
+        A callable implementing the user-defined consume function.
+        The first argument is the context argument of type
+        ScalarAggregateUdfContext.
+        Then, it must take arguments equal to the number of
+        in_types defined.
+        To define a varargs function, pass a callable that takes
+        varargs. The last in_type will be the type of all varargs
+        arguments.
```
Comment (westonpace) [inline python/pyarrow/_compute.pyx:2750]: How do varargs work?  Do I define a `*args`?

File: python/pyarrow/_compute.pyx [hunk 17]
```diff
@@ -2641,3 +2722,200 @@ def register_scalar_function(func, function_name, function_doc, in_types,
...
+    merge_func: callable
+        A callable implementing the user-defined merge function.
+        The first argument is the context argument of type
+        ScalarAggregateUdfContext.
+        Then, the second argument it takes is an state object. 
+        This object holds the state with which the current state
+        must be merged with. The current state can be retrieved from
+        the context object which can be acessed by `context.state`.
+        The state doesn't need to be set in the Python side and it is
+        autonomously handled in the C++ backend. The updated state must
```
Comment (westonpace) [inline python/pyarrow/_compute.pyx:2763]: I'm not sure I understand the sentence that starts with "The state doesn't need to be set..."

File: python/pyarrow/_compute.pyx [hunk 18]
```diff
@@ -2641,3 +2722,200 @@ def register_scalar_function(func, function_name, function_doc, in_types,
...
+        ScalarAggregateUdfContext.
+        Then, it must take arguments equal to the number of
+        in_types defined.
+        To define a varargs function, pass a callable that takes
+        varargs. The last in_type will be the type of all varargs
+        arguments.
+
+        This function returns the updated state after consuming the 
+        received data.
+    merge_func: callable
```
Comment (westonpace) [inline python/pyarrow/_compute.pyx:2754]: The concept of a "merge function" is not going to be obvious to most users.  It is very possible for an engine to be defined that does not have to worry about the concept of a merge.  I think we need to describe some background here for why a merge is needed in the first place.  Something like:

Aggregates may be calculated across many threads in parallel.  Each thread will call the init function once to generate a state for that thread.  Once all values have been consumed then the threads from each state will be merged together to get the final result state.  The merge function should take two states and combine them.

File: python/pyarrow/_compute.pyx [hunk 19]
```diff
@@ -2641,3 +2722,200 @@ def register_scalar_function(func, function_name, function_doc, in_types,
...
+        the context object which can be acessed by `context.state`.
+        The state doesn't need to be set in the Python side and it is
+        autonomously handled in the C++ backend. The updated state must
+        be returned from this function.
+    finalize_func: callable
+        A callable implementing the user-defined finalize function.
+        The first argument is the context argument of type
+        ScalarUdfContext.
+        Using the context argument the state can be extracted and return
+        type must be an array matching the `out_type`.
```
Comment (westonpace) [inline python/pyarrow/_compute.pyx:2770]: In your C++ example the return type is a scalar?

File: python/pyarrow/_compute.pyx [hunk 20]
```diff
@@ -2641,3 +2722,200 @@ def register_scalar_function(func, function_name, function_doc, in_types,
...
+        must be merged with. The current state can be retrieved from
+        the context object which can be acessed by `context.state`.
+        The state doesn't need to be set in the Python side and it is
+        autonomously handled in the C++ backend. The updated state must
+        be returned from this function.
+    finalize_func: callable
+        A callable implementing the user-defined finalize function.
+        The first argument is the context argument of type
+        ScalarUdfContext.
+        Using the context argument the state can be extracted and return
```
Comment (westonpace) [inline python/pyarrow/_compute.pyx:2769]: "the state can be extracted" is not very obvious.  Maybe something like:

The purpose of the finalize function is to transform the state (which is available in the context argument) into an array.  This array will be the final result of the aggregation.

**Review by westonpace — CHANGES_REQUESTED** [review]
The shape of this looks correct.  I believe you are on the right track.  I think I would like to see at least one example of using a custom UDF in an exec plan.  Also, similar to scalar UDFs, I think we will eventually want an example that uses something other than pyarrow (e.g. numpy) to do the computation (at least for the consume function)

**Pull request closed without merging**